### PR TITLE
file-proxy examples

### PR DESCRIPTION
**Pull request by kbrandwijk**
Set of examples around a webtask-based wrapper for the File API, including a bigger authentication example. To follow on the 'practice what you preach', I changed the first example to use graphql-request. The other examples use [request](https://github.com/request/request).

### CODE REVIEW

**Comment by kbrandwijk** [comment]
I have now included the full authenticated file proxy example. The file-proxy example set is now complete.

**Comment by kbrandwijk** [comment]
To follow on the 'practice what you preach', I changed the first example to use graphql-request.

**Comment by marktani** [comment]
Currently reviewing this example, it's great! 🙏 
When I upload a `png` file to the `simple-file-proxy` and visit its `newUrl`, I get the following log in the DevTools:

> Resource interpreted as Document but transferred with MIME type image/png: "https://wt-nilan_marktanner-gmail_com-0.run.webtask.io/simple/cj49x05nvjdeh01735bvg73hx/cj49xo50v03um0168cr63t9r0".

Do you have any insights into this?

File: file-proxy/README.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+This example uses the same GraphQL schema as `simple-file-proxy`. Follow the instructions above if you haven't created a Graphcool project based on this schema.
+
+Next, deploy the webtask for this example to webtask.io by running the following command:
+```sh
+wt create ./multiple-file-proxy.js --name multiple
+```
+Now, try to upload more than one file at once to your endpoint:
+```sh
+curl -X POST '<webtask endpoint url>/__PROJECT_ID__' -F "file1=@example.png;filename=myname.png;file2=@example2.png;filename=testfile.png"
```
Comment (marktani) [inline file-proxy/README.md]: I could not upload multiple files. Instead, I received the following warning:

> Warning: skip unknown form field: file2=@example2.png

And only the *first* file will be uploaded, while being named `testfile.png`.

**Comment by kbrandwijk** [comment]
It might be a header thing. I'll check it when I'm back behind the computer.

File: file-proxy/auth-file-proxy/auth-file-proxy.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        // The uploaded file will be cleaned up by our watcher
+        res.status(403).send();
+        return;
+      }
+
+      // The request to the actual file
+      var resource = request.get(response.data.MyFile.file.url);
+
+      // As soon as we get a response, we copy the headers
+      // Content-length is removed, because the decrypted file does not have the same length
```
Comment (marktani) [inline file-proxy/auth-file-proxy/auth-file-proxy.js]: is it feasible to adjust the `content-length` header with the new length, instead?

**Comment by marktani** [comment]
I have verified the following examples to work:

* `simple-file-proxy`
* `extension-file-proxy`

I found problems with `multiple-file-proxy`, as detailed above.
I read the other examples thoroughly and they look great as well 👍

File: file-proxy/README.md (hunk 1 above)
Comment (kbrandwijk) [inline file-proxy/README.md]: The syntax of the curl command is wrong. It requires to seperate -F sections. I will update the readme.

**Comment by kbrandwijk** [comment]
@marktani, I updated the readme with the correct curl syntax for the multi-file-upload, I fixed the auth-file-proxy content-length issue (by storing/retrieving original file size in `MyFile`).
The warning from Chrome is because I proxy the original request, it's only a warning, and it might be able to remove with additional tweaking, but that's outside of the scope of this example.

File: file-proxy/README.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+```sh
+wt create ./simple-file-proxy.js --name simple
+```
+
+After running this command, `wt-cli` will return the url of your webtask. Using the `-name` argument, you can specify a different name for the webtask (the last part of the url).
+
+Your webtask is now deployed! You can upload a file using any tool that support `multipart/form-data`, like Postman, or the command-line tool `curl`:
+
+```sh
+curl -X POST '<webtask endpoint url>/__PROJECT_ID__' -F "data=@small-logo.png;filename=small.png"
```
Comment (yusinto) [inline file-proxy/README.md]: Nit: use single quotes throughout to be consistent:
curl -X POST '<webtask endpoint url>/__GRAPHCOOL_PROJECT_ID__' -F 'data=@small-logo.png;filename=small.png'

File: file-proxy/simple-file-proxy.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
+"use latest"
```
Comment (yusinto) [inline file-proxy/simple-file-proxy.js]: I have made some minor changes to this file, perhaps it's easier to view these changes directly [here](https://github.com/yusinto/functions/blob/master/file-proxy/simple-file-proxy.js)

**Comment by kbrandwijk** [comment]
@yusinto I processed your code styling suggestions

File: file-proxy/encrypted-file-proxy.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import express from 'express';
+import crypto from 'crypto';
+import multiparty from 'multiparty';
+import formData from 'form-data';
+import request from 'request';
+
+const app = express();
+
+// The upload endpoint
+app.post('/:projectid', (req, res) => {
```
Comment (yusinto) [inline file-proxy/encrypted-file-proxy.js]: I think you mean `projectId`? :)
I also have made some minor adjustments and corrections [here](https://github.com/yusinto/functions/blob/master/file-proxy/encrypted-file-proxy.js)

Comment (kbrandwijk) [inline file-proxy/encrypted-file-proxy.js]: Something is definitely off with that change tracking when you make changes. It's easier to commit them to the PR like @marktani did above. Anyways, the projectId was one of the things I overlooked having to do a manual merge of your previous changes... I will update that. What other things did you change? Only the use of graphql-request in encrypted-file-proxy?

**Comment by yusinto** [comment]
Those two are the biggest ones. I also replaced double quotes with single
to be consistent and changed string concatenation to string templating.

Apologies about the way I use github! I'm reviewing auth-file-proxy and
i'll try and submit a PR for that to make it easier.

**Comment by kbrandwijk** [comment]
My last commit is not showing up here...

**Comment by marktani** [comment]
> PR and history completely screwed, unable to squash. Please squash on PR merge...

😞

**Comment by marktani** [comment]
Thanks so much @kbrandwijk for that amazing contribution! 🙏 Also thanks to @yusinto for the help reviewing this. Next time let's make sure the commits are not messed up 😆

**Pull request merged**